cpp/include/cudf/ast/expressions.hpp

```diff
@@ -46,27 +46,35 @@ struct expression {
  */
 enum class ast_operator {
   // Binary operators
-  ADD,            ///< operator +
-  SUB,            ///< operator -
-  MUL,            ///< operator *
-  DIV,            ///< operator / using common type of lhs and rhs
-  TRUE_DIV,       ///< operator / after promoting type to floating point
-  FLOOR_DIV,      ///< operator / after promoting to 64 bit floating point and then
-                  ///< flooring the result
-  MOD,            ///< operator %
-  PYMOD,          ///< operator % but following python's sign rules for negatives
-  POW,            ///< lhs ^ rhs
-  EQUAL,          ///< operator ==
-  NOT_EQUAL,      ///< operator !=
-  LESS,           ///< operator <
-  GREATER,        ///< operator >
-  LESS_EQUAL,     ///< operator <=
-  GREATER_EQUAL,  ///< operator >=
-  BITWISE_AND,    ///< operator &
-  BITWISE_OR,     ///< operator |
-  BITWISE_XOR,    ///< operator ^
-  LOGICAL_AND,    ///< operator &&
-  LOGICAL_OR,     ///< operator ||
+  ADD,               ///< operator +
+  SUB,               ///< operator -
+  MUL,               ///< operator *
+  DIV,               ///< operator / using common type of lhs and rhs
+  TRUE_DIV,          ///< operator / after promoting type to floating point
+  FLOOR_DIV,         ///< operator / after promoting to 64 bit floating point and then
+                     ///< flooring the result
+  MOD,               ///< operator %
+  PYMOD,             ///< operator % but following python's sign rules for negatives
+  POW,               ///< lhs ^ rhs
+  EQUAL,             ///< operator ==
+  NULL_EQUAL,        ///< NULL_EQUAL(null, null) is true, NULL_EQUAL(null, valid) is false, and
+                     ///< NULL_EQUAL(valid, valid) == EQUAL(valid, valid)
+  NOT_EQUAL,         ///< operator !=
+  LESS,              ///< operator <
+  GREATER,           ///< operator >
+  LESS_EQUAL,        ///< operator <=
+  GREATER_EQUAL,     ///< operator >=
+  BITWISE_AND,       ///< operator &
+  BITWISE_OR,        ///< operator |
+  BITWISE_XOR,       ///< operator ^
+  LOGICAL_AND,       ///< operator &&
+  NULL_LOGICAL_AND,  ///< NULL_LOGICAL_AND(null, null) is null, NULL_LOGICAL_AND(null, true) is
+                     ///< null, NULL_LOGICAL_AND(null, false) is false, and NULL_LOGICAL_AND(valid,
+                     ///< valid) == LOGICAL_AND(valid, valid)
+  LOGICAL_OR,        ///< operator ||
+  NULL_LOGICAL_OR,   ///< NULL_LOGICAL_OR(null, null) is null, NULL_LOGICAL_OR(null, true) is true,
+                     ///< NULL_LOGICAL_OR(null, false) is null, and NULL_LOGICAL_OR(valid, valid) ==
+                     ///< LOGICAL_OR(valid, valid)
```

Comment:
For the `NULL_LOGICAL_AND` / `OR`, the Java comment says "using Spark rules for null inputs" and the C++ comment gives an explicit definition. It would be good to make the comments match. "Spark rules" alone is probably an insufficient description - perhaps including both pieces of information would be best? I don't have a strong opinion except that they should be consistent.

Reply:
Discussed offline: I don't want them to match exactly because the C++ code will be read by libcudf devs who aren't necessarily familiar with Spark rules, unlike our Java developers. However, I've updated these docstrings to explicitly indicate that the behavior described here corresponds to Spark rules, which should help bridge the gap.